openedx/features/effort_estimation/block_transformers.py

```diff
@@ -0,0 +1,225 @@
+    DEFAULT_WPM = 265  # words per minute
+
+    class MissingEstimationData(Exception):
+        pass
+
+    @classmethod
+    def name(cls):
+        """Unique identifier for the transformer's class; same identifier used in setup.py."""
+        return 'effort_estimation'
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Grabs raw estimates for leaf content.
+
+        Pooled estimates higher on verticals and such is done by transform() below, because what each user sees could
+        be different.
+        """
+        block_structure.request_xblock_fields('category')
+        block_structure.request_xblock_fields('global_speed', 'only_on_web')  # video fields
+
+        collection_cache = {}  # collection methods can stuff some temporary data here
+
+        try:
+            for block_key in block_structure.topological_traversal():
+                xblock = block_structure.get_xblock(block_key)
+
+                collections = {
+                    'html': cls._collect_html_effort,
+                    'video': cls._collect_video_effort,
+                }
+
+                if xblock.category in collections:
+                    collections[xblock.category](block_structure, block_key, xblock, collection_cache)
+
+        except cls.MissingEstimationData:
+            # Some bit of required data is missing. Likely some duration info is missing from the video pipeline.
+            # Rather than attempt to work around it, just set a note for ourselves to not show durations for this
+            # course at all. Better no estimate than a misleading estimate.
+            block_structure.set_transformer_data(cls, cls.DISABLE_ESTIMATION, True)
+
+    @classmethod
+    def _collect_html_effort(cls, block_structure, block_key, xblock, _cache):
+        """Records a word count for later reading speed calculations."""
+        try:
+            text = lxml.html.fromstring(xblock.data).text_content() if xblock.data else ''
+        except Exception as exc:  # pylint: disable=broad-except
+            raise cls.MissingEstimationData() from exc
+
+        block_structure.set_transformer_block_field(block_key, cls, cls.HTML_WORD_COUNT, len(text.split()))
+
+    @classmethod
+    def _collect_video_effort(cls, block_structure, block_key, xblock, cache):
+        """Records a duration for later viewing speed calculations."""
+        # Lookup all course video metadata at once rather than piecemeal, for performance reasons
+        if cls.CACHE_VIDEO_DURATIONS not in cache:
+            all_videos, _ = get_videos_for_course(str(block_structure.root_block_usage_key.course_key))
+            cache[cls.CACHE_VIDEO_DURATIONS] = {v['edx_video_id']: v['duration'] for v in all_videos}
+
+        # Check if we have a duration. If not, raise an exception that will stop this transformer from affecting
+        # this course.
+        duration = max(0, cache[cls.CACHE_VIDEO_DURATIONS].get(xblock.edx_video_id, 0))
```

Comment:
what is this returning if it isn't in the cache? wonder if there is a clearer way to write this

Reply:
In the `get(xblock.edx_video_id, 0)` bit, the `0` is the default value for a dictionary lookup. So `0` gets returned if it's not in the cache.

Do you know what about the line is not clear? Is the `max(0, ...)` bit making it harder to read?

Reply:
more of a nit so you can feel free to ignore but you could do something like
`cached_duration = cache[cls.CACHE_VIDEO_DURATIONS].get(xblock.edx_video_id)
duration = cached_duration if cached_duration else 0`